src/loaders/workspace_uploader/workspace_uploader.py

```diff
@@ -0,0 +1,557 @@
+    kbase_collection = getattr(args, loader_common_names.KBASE_COLLECTION_ARG_NAME)
+    source_version = getattr(args, loader_common_names.SOURCE_VER_ARG_NAME)
+    root_dir = getattr(args, loader_common_names.ROOT_DIR_ARG_NAME)
+    token_filepath = args.token_filepath
+    upload_file_ext = args.upload_file_ext
+    workers = args.workers
+    keep_job_dir = args.keep_job_dir
+
+    env = args.env
+    kb_base_url = loader_common_names.KB_BASE_URL_MAP[env]
+
+    if workspace_id <= 0:
+        parser.error(f"workspace_id needs to be > 0")
+    if workers < 1 or workers > cpu_count():
+        parser.error(f"minimum worker is 1 and maximum worker is {cpu_count()}")
+
+    uid = os.getuid()
+    username = os.getlogin()
+
+    job_dir = loader_helper.make_job_dir(root_dir, username)
+    collection_source_dir = loader_helper.make_collection_source_dir(
+        root_dir, loader_common_names.DEFAULT_ENV, kbase_collection, source_version
+    )
+    upload_dir = loader_helper.make_collection_source_dir(root_dir, env, kbase_collection, source_version)
+    output_dir = loader_helper.make_sourcedata_ws_dir(root_dir, env, workspace_id)
+
+    proc = None
+    conf = None
+
+    try:
+        # start podman service
+        proc = loader_helper.start_podman_service(uid)
+
+        # set up conf, start callback server, and upload assemblies to workspace
+        conf = Conf(
+            job_dir,
+            output_dir,
+            kb_base_url,
+            token_filepath,
+            loader_common_names.CALLBACK_UPLOADER_IMAGE_NAME,
+            workers,
+            _process_input,
+        )
+        workspace_name = conf.ws.get_workspace_info({"id": workspace_id})[1]
+
+        count, wait_to_upload_assemblies = _fetch_assemblies_to_upload(
+            env, 
+            workspace_id, 
+            collection_source_dir,
+            upload_file_ext,
+        )
+
+        if not wait_to_upload_assemblies:
+            print(f"All {count} assembly files already exist in workspace {workspace_id}")
+            return
+
+        wtus_len = len(wait_to_upload_assemblies)
+        print(f"Originally planned to upload {count} assembly files")
+        print(f"Detected {count - wtus_len} assembly files already exist in workspace")
+
+        temp_dir = _prepare_skd_job_dir_to_upload(job_dir, wait_to_upload_assemblies)
+        print(f"{wtus_len} assemblies in {temp_dir} are ready to upload to workspace {workspace_id}")
+
+        start = time.time()
+        failed_names = _upload_assembly_files_in_parallel(
+            conf,
+            env,
+            workspace_name,
+            workspace_id,
+            upload_dir,
+            temp_dir,
+            wait_to_upload_assemblies,
+        )
+        
+        assembly_count = wtus_len - len(failed_names)
+        upload_speed = (time.time() - start) / assembly_count
+        print(f"\nSuccessfully upload {assembly_count} assemblies with {workers} workers, average {upload_speed:.2f}s/assembly.")
```

Comment:
It's not totally clear that the parallel workers are taken into account when calculating the upload speed. Maybe something like

```
{workers} workers took {upload_time} minutes to upload {assembly_count} assemblies, averaging {assy_per_min} assemblies per minute
```

Reply:
Updated!

Reply:
:+1: